File: src/main/java/org/goafabric/personservice/persistence/extensions/KafkaPublisher.java
package org.goafabric.personservice.persistence.extensions;

import jakarta.persistence.PostPersist;
import jakarta.persistence.PostRemove;
import jakarta.persistence.PostUpdate;
import org.goafabric.personservice.controller.dto.EventData;
import org.goafabric.personservice.persistence.entity.AddressEo;
import org.goafabric.personservice.persistence.entity.PersonEo;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.aot.hint.MemberCategory;
import org.springframework.aot.hint.annotation.RegisterReflection;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Component;

@RegisterReflection(classes = {EventData.class, PersonEo.class, AddressEo.class} //every type we publish needs to be registered
        , memberCategories = { MemberCategory.INVOKE_DECLARED_METHODS, MemberCategory.INVOKE_DECLARED_CONSTRUCTORS})
@Component
public class KafkaPublisher {

    private final Logger log = LoggerFactory.getLogger(this.getClass());
    private final KafkaTemplate<String, Object> kafkaTemplate;
    private final String kafkaServers;

    public KafkaPublisher(KafkaTemplate<String, Object> kafkaTemplate, @Value("${spring.kafka.bootstrap-servers:}") String kafkaServers) {
        this.kafkaTemplate = kafkaTemplate;
        this.kafkaServers = kafkaServers;
    }

    @PostPersist
    public void afterCreate(Object object)  {
        publish("CREATE", object);
    }

    @PostUpdate
    public void afterUpdate(Object object) {
        publish("UPDATE", object);
    }

    @PostRemove
    public void afterDelete(Object object) {
        publish("DELETE", object);
    }

    private void publish(String operation, Object object) {
       if (kafkaServers.isEmpty()) { return; }

        switch (object) {
            case PersonEo person ->
                    publish("person", person.getId(), operation, object);
            case AddressEo address ->
                    publish("person", address.getId(), operation, object);
            default -> throw new IllegalStateException("Type " + object.getClass());
        }
    }

    private void publish(String type, String key, String operation, Object payload) {
        log.info("publishing event of type {}", type);
        kafkaTemplate.send(type, key, payload);

    }

}

File: src/main/kotlin/org/goafabric/personservice/persistence/extensions/KafkaPublisher.kt
package org.goafabric.personservice.persistence.extensions

import jakarta.persistence.PostPersist
import jakarta.persistence.PostRemove
import jakarta.persistence.PostUpdate
import org.apache.kafka.clients.producer.ProducerRecord
import org.goafabric.personservice.extensions.UserContext
import org.goafabric.personservice.logic.PersonMapper
import org.goafabric.personservice.persistence.entity.AddressEo
import org.goafabric.personservice.persistence.entity.PersonEo
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets
import java.util.function.BiConsumer

@Component
class KafkaPublisher(
    private val kafkaTemplate: KafkaTemplate<String, Any>,
    @param:Value("\${spring.kafka.enabled:false}") private val kafkaEnabled: Boolean,
    private val personMapper: PersonMapper
) {
    private val log: Logger = LoggerFactory.getLogger(this.javaClass)

    private enum class DbOperation {  CREATE, UPDATE, DELETE }

    @PostPersist
    fun afterCreate(`object`: Any) {
        publish(DbOperation.CREATE, `object`)
    }

    @PostUpdate
    fun afterUpdate(`object`: Any) {
        publish(DbOperation.UPDATE, `object`)
    }

    @PostRemove
    fun afterDelete(`object`: Any) {
        publish(DbOperation.DELETE, `object`)
    }

    private fun publish(operation: DbOperation, entity: Any) {
        if (!kafkaEnabled) {
            return
        }

        when (entity) {
            is PersonEo  -> publish("person", entity.id!!, operation, personMapper.map(entity))
            is AddressEo -> publish("address", entity.id!!, operation, personMapper.map(entity))
            else -> error("Type " + entity::class)
        }
    }

    //publish both person and address with the same topic to retain order, put Operation and UserContext to Kafka Headers to prevent EventData Wrapper
    private fun publish(topic: String, key: String, operation: DbOperation, payload: Any) {
        log.info("publishing event of type {}", topic)
        val producerRecord = ProducerRecord(topic, key, payload)
        producerRecord.headers().add("operation", operation.toString().toByteArray(StandardCharsets.UTF_8))

        UserContext.adapterHeaderMap.forEach(BiConsumer { key1: String, value: String ->
            producerRecord.headers().add(
                key1, value.toByteArray(StandardCharsets.UTF_8)
            )
        })

        kafkaTemplate.send(producerRecord)
    }
}
